t/21-needles.t

```diff
@@ -168,4 +178,53 @@ subtest 'handling relative paths in update_needle' => sub {
     };
 };
 
+subtest '_determine_needles_dir_for_job' => sub {
+    my $controller = OpenQA::WebAPI::Controller::Step->new;
+    $controller->app($t->app);
+    $settings{CASEDIR} = 'https://something#fragment';
+    my $sandbox = tempdir(CLEANUP => 1);
+    my $job = $schema->resultset('Jobs')->create_from_settings(\%settings);
+    $job->update({result_dir => $sandbox});
+
+    subtest 'checkout_needles_sha = no' => sub {
+        $controller->app->config->{'scm git'} = {checkout_needles_sha => 'no'};
+        my ($needle_dirs, $needles_var) = $controller->_determine_needles_dir_for_job($job);
+        is($needle_dirs, undef, 'needles directory determined');
+        is($needles_var, undef, 'needles var determined');
+    };
+
+    subtest 'checkout_needles_sha = yes' => sub {
+        $controller->app->config->{'scm git'} = {checkout_needles_sha => 'yes'};
+        my ($needle_dirs, $needles_var) = $controller->_determine_needles_dir_for_job($job);
+        like($needle_dirs, qr|t/data/openqa/share/tests/fedora/needles$|, 'needles dirs is defined');
+        is($needles_var->value, 'https://something#fragment', 'needles var is defined');
+    };
+};
+
+subtest '_create_tmpdir_for_needles_refspec' => sub {
+    my $testresult_dir = setup_fullstack_temp_dir('21-needles');
+    # my $testresult_dir = tempdir("/tmp/$FindBin::Script-XXXX");
+    # path($testresult_dir, 'openqa/db')->make_path->child('db.lock')->spew('');
+    # is($testresult_dir, '', 'test that should fail');
+    # is($ENV{OPENQA_BASEDIR}, '', 'test that should fail #2');
+    # mkdir('t/data/openqa/testresults/00000');
+    my $vars_json_path = path($testresult_dir, '21-needles/openqa/testresults/00000');
+    path($vars_json_path)->make_path;
+    my $vars_json = path($vars_json_path, 'vars.json');
+    my $vars_json_content = '{"NEEDLES_GIT_HASH": "0123456789abcdef0123456789abcdef01234567"}';
+    $vars_json->spew($vars_json_content);
+
+    my $controller = OpenQA::WebAPI::Controller::Step->new;
+    $controller->app($t->app);
+    $settings{CASEDIR} = 'https://something#fragment';
+    $job = $schema->resultset('Jobs')->create_from_settings(\%settings);
+    $job->update({result_dir => $testresult_dir});
```

Comment:
We discussed that you should rather set OPENQA_BASEDIR to avoid messing with the source git dir at all.
And `setup_fullstack_temp_dir` does that automatically. And also the `db.lock` thing.
https://github.com/os-autoinst/openQA/blob/master/t/lib/OpenQA/Test/Utils.pm#L369
And it does not return the testresult dir, but the tempdir it created.

Why is it necessary to change the dir of only one certain job?

Reply:
If result_dir is not setted the subroutine when call `_determine_needles_dir_for_job` will return undef because `realpath($job->needle_dir)` is undef. Do I need to mock needle_dir?